core/templates/expressions/expression-syntax-tree.service.ts

```diff
@@ -23,54 +23,83 @@ import { AppConstants } from 'app.constants';
 import { ExpressionParserService } from
   'expressions/expression-parser.service.ts';
 
-// Exceptions that can be thrown from the evaluation of expressions.
-const ExpressionError = function(): void {};
-ExpressionError.prototype = new Error();
-ExpressionError.prototype.constructor = ExpressionError;
-
-const ExprUndefinedVarError = function(varname: string,
-    envs: Array<object>): void {
-  this.varname = varname;
-  this.envs = envs;
-};
-ExprUndefinedVarError.prototype = new ExpressionError();
-ExprUndefinedVarError.prototype.constructor = ExprUndefinedVarError;
-ExprUndefinedVarError.prototype.name = 'ExprUndefinedVarError';
-ExprUndefinedVarError.prototype.toString = function(): string {
-  return this.name + ': ' + this.varname + ' not found in ' + this.envs;
-};
-
-const ExprWrongNumArgsError = function(args: Array<number|string>,
-    expectedMin: number, expectedMax: number): void {
-  this.args = args;
-  this.expectedMin = expectedMin;
-  this.expectedMax = expectedMax;
-};
-ExprWrongNumArgsError.prototype = new ExpressionError();
-ExprWrongNumArgsError.prototype.constructor = ExprWrongNumArgsError;
-ExprWrongNumArgsError.prototype.name = 'ExprWrongNumArgsError';
-ExprWrongNumArgsError.prototype.toString = function(): string {
-  return this.name + ': {' + this.args + '} not in range [' +
+class ExpressionError extends Error {
+  constructor() {
+    super();
+    // NOTE TO DEVELOPERS: As per the recommendation of TypeScript
+    //  we need to adjust the prototype manually right after the super() call
+    //  For more information refer to this link below:
+    //  https://github.com/Microsoft/TypeScript-wiki/blob/master/Breaking-Changes.md#extending-built-ins-like-error-array-and-map-may-no-longer-work
+    Object.setPrototypeOf(this, ExpressionError.prototype);
+  }
+}
+
+class ExprUndefinedVarError extends ExpressionError {
+  constructor(public varname: string, public envs: Array<object>) {
```

Comment:
The body is redundant, please make it empty. Adding "public/private" to the parameters in constructors does the assignments automagically for us.

Reply:
NVM, didn't notice you were extending `Error`.